kernel/seccomp.c

```diff
@@ -1115,6 +1115,22 @@ static int seccomp_do_user_notification(int this_syscall,
 			mutex_unlock(&match->notify_lock);
 			goto wait;
 		}
+
+		/*
+		 * Check if we were woken up by addfd and the notification is
+		 * replied, in which case just allocate the fd and return it.
+		 */
+		if (addfd && n.state == SECCOMP_NOTIFY_REPLIED) {
+			seccomp_handle_addfd(addfd);
+			n.val = addfd->ret;
+
+			/* If we fail to allocate an fd mark this as err */
+			if (n.val < 0) {
+				n.error = n.val;
+				n.val = 0;
```

Comment:
Why 0?

Can we add a unit test to check that the return value and the errno of the syscall is correct in this case?

I assume it should return:
- return value: -1
- errno: EMFILE ("The per-process limit on the number of open file descriptors has been reached")

Reply:
Yeah, if you set error to -errno, the call to `syscall_set_return_value()` at the end of this function will handle that (return -1 and set errno to the target). Here is the function doc: https://github.com/torvalds/linux/blob/19c329f6808995b142b3966301f217c831e7cf31/include/asm-generic/syscall.h#L85-L100

If you grep the kernel tree for usage of `syscall_set_return_value()` you will find that it is always used like this (error set to -errno and value is 0). I did the same for consistency, and tested and this works fine.

However, I noticed that if I don't set these values here, the proper thing is returned to the target if I add more fds that we can: -1 and errno set to too many files. I didn't fully understand if that was happening by chance or it was on purpose (as in that case we call `syscall_set_return_value()` with `err=0` and `val=-errno`). To be safe I used the same convention it is used in the rest of the kernel code, as also the documentation of `syscall_set_return_value()` doesn't say that should work either. But planned to ask in the cover letter.

Reply:
Regarding the test, I wanted to ask if this looks good and I can add more tests in follow-up patches. What do you think? I wanted to avoid spending more time on tests now, before knowing the overall impression of the patch

Reply:
Fine for me to have this first iteration. You could list in the cover letter the list of missing tests, and mark this patch set as RFC.